File: python/src/a3_builder.py
"""
The intent is to separate the construction of a complex object from its representation,
so that the same construction process can create different representations.

Key participants:
- Builder: Specifies an abstract interface for creating parts of a Product
- ConcreteBuilder: Constructs and assembles parts of the product by implementing Builder
- Director: Constructs an object using the Builder interface
- Product: The object under construction

Key benefits:
- Lets you vary a product's internal representation
- Isolates construction code from representation code
- Gives you finer control over the construction process

Real-world example: Building an HTTP request with various options (method, URL, body,
headers, timeout) step by step.
"""
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Optional


# ----- Product -----
@dataclass
class HttpRequest:
    method: str = "GET"
    url: str = ""
    body: str = ""
    headers: dict[str, str] = field(default_factory=dict)
    timeout: int = 5000

    def __str__(self) -> str:
        headers_str = ", ".join(f"{k}: {v}" for k, v in self.headers.items())
        return (f"HttpRequest(method={self.method}, url={self.url}, "
                f"timeout={self.timeout}ms, headers=[{headers_str}])")


# ----- Builder -----
class HttpRequestBuilder(ABC):
    """Abstract Builder"""

    @abstractmethod
    def set_method(self, method: str) -> "HttpRequestBuilder":
        pass

    @abstractmethod
    def set_url(self, url: str) -> "HttpRequestBuilder":
        pass

    @abstractmethod
    def set_body(self, body: str) -> "HttpRequestBuilder":
        pass

    @abstractmethod
    def set_timeout(self, timeout: int) -> "HttpRequestBuilder":
        pass

    @abstractmethod
    def add_header(self, key: str, value: str) -> "HttpRequestBuilder":
        pass

    @abstractmethod
    def build(self) -> HttpRequest:
        pass


# ----- Concrete Builder -----
class ConcreteHttpRequestBuilder(HttpRequestBuilder):
    """Concrete Builder that constructs HTTP requests"""

    def __init__(self) -> None:
        self._request = HttpRequest()

    def set_method(self, method: str) -> "HttpRequestBuilder":
        self._request.method = method
        return self

    def set_url(self, url: str) -> "HttpRequestBuilder":
        self._request.url = url
        return self

    def set_body(self, body: str) -> "HttpRequestBuilder":
        self._request.body = body
        return self

    def set_timeout(self, timeout: int) -> "HttpRequestBuilder":
        self._request.timeout = timeout
        return self

    def add_header(self, key: str, value: str) -> "HttpRequestBuilder":
        self._request.headers[key] = value
        return self

    def build(self) -> HttpRequest:
        # Validate before building
        if not self._request.url:
            raise ValueError("URL is required")
        return self._request


# ----- Director (optional) -----
class HttpRequestDirector:
    """
    Director class that defines the construction process using the Builder.
    This is optional - clients can also use the builder directly.
    """

    def __init__(self, builder: HttpRequestBuilder) -> None:
        self._builder = builder

    def construct_post_request(self, url: str, body: str) -> HttpRequest:
        """Constructs a typical POST request"""
        return (self._builder
                .set_method("POST")
                .set_url(url)
                .set_body(body)
                .add_header("Content-Type", "application/json")
                .set_timeout(5000)
                .build())

    def construct_get_request(self, url: str) -> HttpRequest:
        """Constructs a typical GET request"""
        return (self._builder
                .set_method("GET")
                .set_url(url)
                .set_timeout(3000)
                .build())


# ----- Example usage -----
def builder_example() -> None:
    """Demonstrates the Builder pattern with method chaining"""
    print("\n--- Builder Pattern: HTTP Request ---\n")

    # Example 1: Using builder directly with method chaining
    print("1. Building custom request with method chaining:")
    request = (ConcreteHttpRequestBuilder()
               .set_method("POST")
               .set_url("https://example.com/api")
               .set_body('{"id": 42}')
               .add_header("Authorization", "Bearer token123")
               .set_timeout(5000)
               .build())

    print(f"   {request}")

    # Example 2: Using Director for common patterns
    print("\n2. Using Director for predefined patterns:")
    builder = ConcreteHttpRequestBuilder()
    director = HttpRequestDirector(builder)

    post_request = director.construct_post_request(
        "https://api.example.com/users",
        '{"name": "John", "email": "john@example.com"}'
    )
    print(f"   POST: {post_request}")

    # Reset and build GET request
    get_request = director.construct_get_request("https://api.example.com/users/42")
    print(f"   GET:  {get_request}")


if __name__ == "__main__":
    builder_example()

File: rust/src/a3_builder.rs
//! # Builder Pattern
//!
//! The Builder pattern separates the construction of complex objects from
//! their representation, allowing the same construction process to create
//! different representations.
//!
//! ## Key Participants
//! - **Builder**: Specifies abstract interface for building parts
//! - **ConcreteBuilder**: Constructs and assembles parts
//! - **Director**: Constructs object using Builder interface
//! - **Product**: The object under construction
//!
//! ## Benefits
//! - Lets you vary product's internal representation
//! - Isolates construction code from representation code
//! - Gives finer control over construction process

#[derive(Debug, Clone)]
pub struct HttpRequest {
    pub method: String,
    pub url: String,
    pub headers: Vec<(String, String)>,
    pub body: Option<String>,
    pub timeout_ms: u64,
}

impl HttpRequest {
    pub fn send(&self) {
        println!("Sending HTTP Request:");
        println!("  {} {}", self.method, self.url);
        println!("  Headers: {} entries", self.headers.len());
        if let Some(body) = &self.body {
            println!("  Body: {} bytes", body.len());
        } else {
            println!("  Body: (none)");
        }
        println!("  Timeout: {}ms", self.timeout_ms);
    }
}

pub struct HttpRequestBuilder {
    method: Option<String>,
    url: Option<String>,
    headers: Vec<(String, String)>,
    body: Option<String>,
    timeout_ms: Option<u64>,
}

impl HttpRequestBuilder {
    pub fn new() -> Self {
        Self {
            method: None,
            url: None,
            headers: Vec::new(),
            body: None,
            timeout_ms: None,
        }
    }

    pub fn method(mut self, method: &str) -> Self {
        self.method = Some(method.to_string());
        self
    }

    pub fn url(mut self, url: &str) -> Self {
        self.url = Some(url.to_string());
        self
    }

    pub fn header(mut self, key: &str, value: &str) -> Self {
        self.headers.push((key.to_string(), value.to_string()));
        self
    }

    pub fn body(mut self, body: &str) -> Self {
        self.body = Some(body.to_string());
        self
    }

    pub fn timeout(mut self, timeout_ms: u64) -> Self {
        self.timeout_ms = Some(timeout_ms);
        self
    }

    pub fn build(self) -> Result<HttpRequest, String> {
        let method = self.method.ok_or("Method is required")?;
        let url = self.url.ok_or("URL is required")?;
        let timeout_ms = self.timeout_ms.unwrap_or(5000);

        Ok(HttpRequest {
            method,
            url,
            headers: self.headers,
            body: self.body,
            timeout_ms,
        })
    }
}

impl Default for HttpRequestBuilder {
    fn default() -> Self {
        Self::new()
    }
}

/// Example demonstrating Builder pattern
pub fn example() {
    println!("\n--- Builder Pattern Example ---\n");

    // Simple GET request
    println!("1. Simple GET Request:");
    let get_request = HttpRequestBuilder::new()
        .method("GET")
        .url("https://api.example.com/users")
        .timeout(3000)
        .build()
        .unwrap();
    get_request.send();

    // Complex POST request
    println!("\n2. Complex POST Request:");
    let post_request = HttpRequestBuilder::new()
        .method("POST")
        .url("https://api.example.com/data")
        .header("Content-Type", "application/json")
        .header("Authorization", "Bearer token123")
        .body(r#"{"id": 42, "name": "John Doe"}"#)
        .timeout(10000)
        .build()
        .unwrap();
    post_request.send();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_builder_creates_request() {
        let request = HttpRequestBuilder::new()
            .method("GET")
            .url("https://example.com")
            .build()
            .unwrap();

        assert_eq!(request.method, "GET");
        assert_eq!(request.url, "https://example.com");
    }

    #[test]
    fn test_builder_with_all_fields() {
        let request = HttpRequestBuilder::new()
            .method("POST")
            .url("https://example.com")
            .header("X-Custom", "value")
            .body("test body")
            .timeout(5000)
            .build()
            .unwrap();

        assert_eq!(request.method, "POST");
        assert_eq!(request.headers.len(), 1);
        assert_eq!(request.body, Some("test body".to_string()));
        assert_eq!(request.timeout_ms, 5000);
    }

    #[test]
    fn test_builder_requires_method() {
        let result = HttpRequestBuilder::new()
            .url("https://example.com")
            .build();

        assert!(result.is_err());
        assert_eq!(result.unwrap_err(), "Method is required");
    }

    #[test]
    fn test_builder_requires_url() {
        let result = HttpRequestBuilder::new()
            .method("GET")
            .build();

        assert!(result.is_err());
        assert_eq!(result.unwrap_err(), "URL is required");
    }

    #[test]
    fn test_builder_default_timeout() {
        let request = HttpRequestBuilder::new()
            .method("GET")
            .url("https://example.com")
            .build()
            .unwrap();

        assert_eq!(request.timeout_ms, 5000); // Default timeout
    }
}
